sdk/src/azure/core/az_mqtt_connection_hfsm.c

```diff
@@ -0,0 +1,384 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <azure/core/az_mqtt.h>
+#include <azure/core/az_mqtt_connection.h>
+#include <azure/core/az_platform.h>
+#include <azure/core/az_result.h>
+#include <azure/core/internal/az_log_internal.h>
+
+#include <azure/core/_az_cfg.h>
+
+static az_result root(az_event_policy* me, az_event event);
+
+static az_result idle(az_event_policy* me, az_event event);
+static az_result started(az_event_policy* me, az_event event);
+static az_result faulted(az_event_policy* me, az_event event);
+
+static az_result connecting(az_event_policy* me, az_event event);
+static az_result connected(az_event_policy* me, az_event event);
+static az_result disconnecting(az_event_policy* me, az_event event);
+static az_result reconnect_timeout(az_event_policy* me, az_event event);
+
+static az_event_policy_handler _get_parent(az_event_policy_handler child_state)
+{
+  az_event_policy_handler parent_state;
+
+  if (child_state == root)
+  {
+    parent_state = NULL;
+  }
+  else if (child_state == idle || child_state == started || child_state == faulted)
+  {
+    parent_state = root;
+  }
+  else if (
+      child_state == connecting || child_state == connected || child_state == disconnecting
+      || child_state == reconnect_timeout)
+  {
+    parent_state = started;
+  }
+  else
+  {
+    // Unknown state.
+    az_platform_critical_error();
+    parent_state = NULL;
+  }
+
+  return parent_state;
+}
+
+static az_result root(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+
+  if (_az_LOG_SHOULD_WRITE(event.type))
```

Comment:
Probably not the place for this comment, but if we're passing the event.type into _az_LOG_WRITE, why isn't should_write checked within there?

Reply:
This is following logging best practices for az_sdk (and other SDKs).
The main reason is to avoid extra computation for logging when there is no listener attached (this is not the case but imagine that a function call is made to compute one of the logging parameters).
The second reason is that if the compiler can determine that no listener is ever attached, this code should be pruned.